File: lib/radiosonde.py
"""Radiosonde station and sounding fetcher — mirrors lib/radiosonde.mjs."""

from __future__ import annotations
import json
import math
import urllib.request
from pathlib import Path
from typing import Any

_CACHE_PATH = Path.home() / ".cache" / "windy-radiosonde" / "stations.json"
_SNAPSHOT_PATH = Path(__file__).parent.parent / "stations-example.json"

_BASE_OBS = "https://node.windy.com/obs/measurement/v2/radiosonde"
_BASE_DL = "https://dl.windy.com/obs/measurement/v2/radiosonde"

_stations: list[dict] | None = None


def _haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    R = 6371
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = math.sin(dlat / 2) ** 2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlon / 2) ** 2
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def _fetch(url: str) -> Any:
    with urllib.request.urlopen(url) as r:
        if r.status != 200:
            raise RuntimeError(f"HTTP {r.status} from {url}")
        return json.loads(r.read())


def _load_stations() -> list[dict]:
    global _stations
    if _stations is not None:
        return _stations
    for path in (_CACHE_PATH, _SNAPSHOT_PATH):
        try:
            raw = json.loads(path.read_text())
            _stations = [{"id": k, **v} for k, v in raw.items()]
            return _stations
        except OSError:
            continue
    raise RuntimeError("No station data found (cache or snapshot)")


def find_stations(
    *,
    name: str | None = None,
    station_id: str | None = None,
    lat: float | None = None,
    lon: float | None = None,
    radius_km: float = 500,
    limit: int = 10,
) -> list[dict]:
    """Find stations by name substring, station ID, or lat/lon proximity."""
    stations = _load_stations()

    if station_id is not None:
        return [s for s in stations if s["id"] == station_id]

    if name is not None:
        q = name.lower()
        scored = []
        for s in stations:
            n = (s.get("name") or "").lower()
            icao = (s.get("icao") or "").lower()
            if n == q or icao == q:
                score = 3
            elif n.startswith(q) or icao.startswith(q):
                score = 2
            elif q in n or q in icao:
                score = 1
            else:
                continue
            scored.append((score, s))
        scored.sort(key=lambda x: -x[0])
        return [s for _, s in scored[:10]]

    if lat is not None and lon is not None:
        with_dist = []
        for s in stations:
            d = round(_haversine_km(lat, lon, s["lat"], s["lon"]), 1)
            if d <= radius_km:
                with_dist.append({**s, "distance_km": d})
        with_dist.sort(key=lambda s: s["distance_km"])
        return with_dist[:limit]

    raise ValueError("provide lat+lon, station_id, or name")


def list_ascents(station_id: str) -> list[dict]:
    """Return ascents for a station (~14 days), newest first."""
    from datetime import datetime, timezone
    data = _fetch(f"{_BASE_OBS}/{station_id}/1/1?pr=0&sc=0&token2=pending")
    ascents = [
        {
            "time_ms": h["time"],
            "time_utc": datetime.fromtimestamp(h["time"] / 1000, tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.000Z"),
            "format": h["format"],
        }
        for h in data.get("history", [])
    ]
    ascents.sort(key=lambda a: -a["time_ms"])
    return ascents


def get_sounding(station_id: str, time_ms: int | None = None) -> dict:
    """Fetch a sounding as a GeoJSON FeatureCollection.

    If time_ms is omitted, fetches the latest available ascent.
    Point features contain: pressure, temp (K), dewpoint (K), wind_u, wind_v, gpheight.
    Last feature is a LineString (flight path).
    """
    if time_ms is None:
        ascents = list_ascents(station_id)
        if not ascents:
            raise RuntimeError(f"No ascents available for {station_id}")
        time_ms = ascents[0]["time_ms"]
    return _fetch(f"{_BASE_DL}/{station_id}/download?time={time_ms}&format=fm94")

File: lib/radiosonde.mjs
import { readFile } from 'fs/promises';
import { homedir } from 'os';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const CACHE_PATH = join(homedir(), '.cache', 'windy-radiosonde', 'stations.json');
const SNAPSHOT_PATH = join(dirname(fileURLToPath(import.meta.url)), '..', 'stations-example.json');

const BASE_OBS = 'https://node.windy.com/obs/measurement/v2/radiosonde';
const BASE_DL = 'https://dl.windy.com/obs/measurement/v2/radiosonde';

function haversineKm(lat1, lon1, lat2, lon2) {
  const R = 6371;
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a = Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * Math.sin(dLon / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

async function windyFetch(url) {
  const r = await fetch(url);
  if (!r.ok) throw new Error(`HTTP ${r.status} from ${url}`);
  return r.json();
}

let _stations = null;

async function loadStations() {
  if (_stations) return _stations;
  let raw;
  try {
    raw = JSON.parse(await readFile(CACHE_PATH, 'utf8'));
  } catch {
    raw = JSON.parse(await readFile(SNAPSHOT_PATH, 'utf8'));
  }
  _stations = Object.entries(raw).map(([id, s]) => ({ id, ...s }));
  return _stations;
}

/**
 * Find radiosonde stations.
 *
 * Input variants (exactly one required):
 *   { lat, lon, radiusKm?, limit? }  — proximity search
 *   { stationId }                    — direct lookup by ID
 *   { name }                         — substring search on name/ICAO, best match first
 */
export async function findStations(input) {
  const stations = await loadStations();

  if (input.stationId != null) {
    const s = stations.find(s => s.id === input.stationId);
    return s ? [s] : [];
  }

  if (input.name != null) {
    const q = input.name.toLowerCase();
    const scored = stations
      .map(s => {
        const n = (s.name || '').toLowerCase();
        const icao = (s.icao || '').toLowerCase();
        let score = 0;
        if (n === q || icao === q) score = 3;
        else if (n.startsWith(q) || icao.startsWith(q)) score = 2;
        else if (n.includes(q) || icao.includes(q)) score = 1;
        return { s, score };
      })
      .filter(x => x.score > 0)
      .sort((a, b) => b.score - a.score);
    return scored.slice(0, 10).map(x => x.s);
  }

  if (input.lat != null && input.lon != null) {
    const radiusKm = input.radiusKm ?? 500;
    const limit = input.limit ?? 10;
    return stations
      .map(s => ({ ...s, distance_km: Math.round(haversineKm(input.lat, input.lon, s.lat, s.lon) * 10) / 10 }))
      .filter(s => s.distance_km <= radiusKm)
      .sort((a, b) => a.distance_km - b.distance_km)
      .slice(0, limit);
  }

  throw new Error('findStations: provide lat+lon, stationId, or name');
}

/**
 * List available ascents for a station (~14 days history).
 * Returns array of { time_ms, time_utc, format } sorted newest first.
 */
export async function listAscents(stationId) {
  const data = await windyFetch(`${BASE_OBS}/${stationId}/1/1?pr=0&sc=0&token2=pending`);
  return (data.history || [])
    .map(h => ({ time_ms: h.time, time_utc: new Date(h.time).toISOString(), format: h.format }))
    .sort((a, b) => b.time_ms - a.time_ms);
}

/**
 * Fetch a sounding as a raw GeoJSON FeatureCollection.
 * If timeMs is omitted, fetches the latest available ascent.
 * Point features contain: pressure, temp, dewpoint, wind_u, wind_v, gpheight.
 * Last feature is a LineString (flight path) — filter with f.geometry.type === 'Point' if unwanted.
 */
export async function getSounding(stationId, timeMs) {
  if (timeMs == null) {
    const ascents = await listAscents(stationId);
    if (!ascents.length) throw new Error(`No ascents available for ${stationId}`);
    timeMs = ascents[0].time_ms;
  }
  return windyFetch(`${BASE_DL}/${stationId}/download?time=${timeMs}&format=fm94`);
}
